# Kickstarter and SQL `10 points`

Since I claimed that the point of Lede was to be able to learn anything using Google, I should probably test that, too, yeah?

Working with **SQL databases** is something we didn't learn! It comes in a lot of formats - PostreSQL, MySQL, Sqlite - but they are all generally the same thing.

SQL is great for when you're reading in big big big datasets: instead of waiting an hour for `.read_csv` to crash on a giant file that you're going to filter anyway, instead you just ask SQL to give you a subset of the data to work on.

**Topics**

* Learning new things with the power of the internet
* SQL

### The data

**Kickstarter** is a website people use to raise money for projects. It used to be really popular, but it's since died down a bit. We're going to work with [a dataset about Kickstarter in SQLite format](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/EOYBXM).

You'll want to download the database by going to **Access Dataset** and then **Download ZIP**.

It's a big file, and it's compressed! We'll be using the `kickstarter.db` file that is inside the zip file: when you open up the zip, I extracted it using [The Unarchiver](https://theunarchiver.com/) by telling it to open up the `.z01` file (it's so big it's spread across both the `.zip` and the `.z01`). 

## Open the dataset

Just like an Excel file can have different sheets, a SQL database has different tables.

In [2]:
import pandas as pd

pd.set_option("display.max_columns", None)
pd.set_option('display.float_format', '{:,.2f}'.format)

SQLite + pandas doesn't use the "normal" way to ask for tables, so I'll give that line to you.

In [3]:
import sqlite3

# Open a connection
conn = sqlite3.connect("files/kickstarter.db")

# Make a query, save it into a dataframe
df = pd.read_sql_query('SELECT name FROM sqlite_master WHERE type = "table";', conn)

# Close the connection
conn.close()

# See the results in the dataframe
df

name
0         urls_to_scrape
1              all_files
2               category
3                creator
4          funding_trend
5                   item
6             livestream
7               location
8                project
9                 reward
10                update
11              comments
12  creator_external_url
13           reward_item
14           creator_bio

## Simple selects `3 points`

### Select all of the data for the first 4 kickstarter projects

They live in the `project` table. This will help you understand how SQL queries work.

In [12]:
def dbquery(query):
    conn = sqlite3.connect("files/kickstarter.db")
    df = pd.read_sql_query(query, conn)
    conn.close()
    return df

In [13]:
dbquery('SELECT * FROM project ORDER BY created_at ASC LIMIT 4;')

id       state                                        url_project  \
0  1860890148      failed  https://www.kickstarter.com/projects/perry/gra...   
1  1473103893  successful  https://www.kickstarter.com/projects/williambr...   
2   709707365      failed  https://www.kickstarter.com/projects/videothin...   
3   405307432  successful  https://www.kickstarter.com/projects/Samplefre...   

      url_project_short                                               name  \
0  http://kck.st/aRzK5V  Grace Jones Does Not Give A F$#% T-Shirt (limi...   
1  http://kck.st/caFA3s           TELEVISION LANDSCAPE: Making the Record.   
2  http://kck.st/dymfMD                     CRYSTAL ANTLERS UNTITLED MOVIE   
3  http://kck.st/cO8V5u  Electronola - An electronic gumbo of New Orlea...   

  country  creator_id   location_id  category_id  created_at    deadline  \
0      US           4 12,589,335.00            9  1240333219  1243753199   
1      US  1505954783           NaN           40  1240335335  1253030340   
2      US  1292964752  2,442,047.00           32  1240356396  1248073199   
3      US          11  2,458,833.00           38  1240366270  1244185140   

   updated_at  state_changed_at    successful_at  launched_at   goal  pledged  \
0  1397849643        1243753227              NaN   1240340568   1000   625.00   
1  1479099902        1253030408 1,253,030,408.00   1245374186   2000 2,000.00   
2  1397849644        1248073228              NaN   1240438073  80000    22.00   
3  1397849645        1244185224 1,244,185,224.00   1241192665   4000 4,100.60   

  currency currency_symbol  usd_pledged  static_usd_rate  backers_count  \
0      USD               $       625.00                1             30   
1      USD               $     2,000.00                1             33   
2      USD               $        22.00                1              3   
3      USD               $     4,100.60                1            114   

   comments_count  updates_count spotlight staff_pick  \
0               9              2     False      False   
1               0              0      True      False   
2               1              0     False      False   
3              30             50      True       True   

                                               blurb currency_trailing_code  \
0  About two years ago I realized a profound trut...                   True   
1  William Brittelle's "Television Landscape" is ...                   True   
2  Above is a music video edited together with sh...                   True   
3  UPDATE: Shannon Powell, Walter Payton, Lucien ...                   True   

  disable_communication                                          photo_url  \
0                 False  https://ksr-ugc.imgix.net/assets/011/257/331/e...   
1                 False  https://ksr-ugc.imgix.net/assets/011/257/333/2...   
2                 False  https://ksr-ugc.imgix.net/assets/011/257/334/1...   
3                 False  https://ksr-ugc.imgix.net/assets/011/257/336/a...   

  profile_background_color  profile_background_image_opacity profile_blurb  \
0                     None                              0.80          None   
1                     None                              0.80          None   
2                     None                              0.80          None   
3                     None                              0.80          None   

   profile_id profile_link_background_color profile_link_text  \
0          59                          None              None   
1          61                          None              None   
2          62                          None              None   
3          63                          None              None   

  profile_link_text_color profile_link_url profile_name  profile_project_id  \
0                    None             None         None                  59   
1                    None             None         None                  61   
2                    None    

### Select the top 10 projects with the highest amount pledged

I only want to see the name and the amount pledged. **The number one result should be the Pebble watch** - if your result is something Japanese, try querying all of the columns and figure out what went wrong.

In [15]:
dbquery('SELECT name,usd_pledged FROM project ORDER BY usd_pledged DESC LIMIT 10;')

name   usd_pledged
0   Pebble Time - Awesome Smartwatch, No Compromises 20,338,986.27
1  COOLEST COOLER: 21st Century Cooler that's Act... 13,285,226.36
2             Pebble 2, Time 2 + All-New Pebble Core 12,779,843.49
3                         Kingdom Death: Monster 1.5 12,393,139.69
4       Pebble: E-Paper Watch for iPhone and Android 10,266,845.74
5  The World's Best TRAVEL JACKET with 15 Feature...  9,192,055.66
6                                  Exploding Kittens  8,782,571.99
7             OUYA: A New Kind of Video Game Console  8,596,474.58
8  THE 7th CONTINENT – What Goes Up, Must Come Down.  7,072,757.00
9             The Everyday Backpack, Tote, and Sling  6,565,782.50

## Filtering your queries `3 points`

### Find me – yes, me, Soma! – in the project creators table

In [16]:
dbquery('SELECT * FROM creator WHERE name = "Jonathan Soma";')

id           name  slug  \
0  546599331  Jonathan Soma  soma   

                                        avatar_thumb  \
0  https://ksr-ugc.imgix.net/assets/005/801/114/d...   

                                        avatar_small  \
0  https://ksr-ugc.imgix.net/assets/005/801/114/d...   

                              urls_web_user  \
0  https://www.kickstarter.com/profile/soma   

                                       avatar_medium is_registered  \
0  https://ksr-ugc.imgix.net/assets/005/801/114/d...          None   

                                       urls_api_user chosen_currency  
0  https://api.kickstarter.com/v1/users/546599331...            None

### Find all of my projects

In [17]:
dbquery('SELECT * FROM project WHERE creator_id = "546599331";')

id       state                                        url_project  \
0   320516934  successful  https://www.kickstarter.com/projects/soma/broo...   
1  1052891075  successful  https://www.kickstarter.com/projects/soma/cook...   
2  1606179513  successful  https://www.kickstarter.com/projects/soma/ice-...   

      url_project_short                                             name  \
0  http://kck.st/d9d9Sx                  Brooklyn Brainery Needs a Home!   
1  http://kck.st/aZ8hta  Cooking Class in a Box [Exotic Cuisine Edition]   
2  http://kck.st/cgpgNy                                   Ice Cream Club   

  country  creator_id  location_id  category_id  created_at    deadline  \
0      US   546599331     12589335            1  1270872639  1278129540   
1      US   546599331     12589335           10  1276719823  1278991380   
2      US   546599331      2459115           10  1271981825  1274055420   

   updated_at  state_changed_at  successful_at  launched_at  goal  pledged  \
0  1397755136        1278129615     1278129615   1271182430  9500 9,629.95   
1  1397756801        1278991807     1278991807   1276792614   200 1,820.44   
2  1397755379        1274056206     1274056206   1272034766   220   485.98   

  currency currency_symbol  usd_pledged  static_usd_rate  backers_count  \
0      USD               $     9,629.95                1            204   
1      USD               $     1,820.44                1             53   
2      USD               $       485.98                1             27   

   comments_count  updates_count spotlight staff_pick  \
0               9              9      True       True   
1               8              7      True       True   
2               7              5      True       True   

                                               blurb currency_trailing_code  \
0  We host awesomely cheap, collaborative classes...                   True   
1  I'll build you a cooking class on an exotic cu...                   True   
2  When it's 100 degrees outside you'll wish you ...                   True   

  disable_communication                                          photo_url  \
0                 False  https://ksr-ugc.imgix.net/assets/011/261/121/e...   
1                 False  https://ksr-ugc.imgix.net/assets/011/263/591/4...   
2                 False  https://ksr-ugc.imgix.net/assets/011/261/487/c...   

  profile_background_color  profile_background_image_opacity profile_blurb  \
0                     None                              0.80          None   
1                     None                              0.80          None   
2                     None                              0.80          None   

   profile_id profile_link_background_color profile_link_text  \
0        4894                          None              None   
1        7857                          None              None   
2        5368                          None              None   

  profile_link_text_color profile_link_url profile_name  profile_project_id  \
0                    None             None         None                4894   
1                    None             None         None                7857   
2                    None             None         None                5368   

  profile_should_show_feature_image_section profile_show_feature_image  \
0                                      True                      False   
1                                      True                      False   
2                                      True                      False   

  profile_state  profile_state_changed_at profile_text_color  \
0      inactive                1425915800               None   
1      inactive                1425915800               None   
2      inactive                1425915800               None   

                                            slug  \
0                 brooklyn-brainery-needs-a-home   
1  cooking-class-in-a-box-exotic-cuisine-edition   
2         

## Calculations `3 points`

### Find how much money has been raised on Kickstarter by **successful** projects.

In [21]:
dbquery('SELECT SUM(usd_pledged) AS total FROM project WHERE state = "successful";')

total
0 3,643,252,201.72

### Find how much money has been raised on Kickstarter by "successful" projects in **the United States**.

The `state` can be successful, failed, live, etc.

In [22]:
dbquery('SELECT SUM(usd_pledged) AS total FROM project WHERE state = "successful" AND country = "US";')

total
0 2,961,700,473.92

## Aggregating `3 points`

## How many projects were successful/unsuccessful/etc?

**Use your SQL query to calculate the answer.** Your result should look something like this:

| |...|...|
|---|---|---|
|0|canceled|36659|
|1|failed|211783|
|2|live|3163|
|3|purged|202|
|4|successful|155130|
|5|suspended|1700|

In [24]:
dbquery('SELECT state, COUNT(*) AS total FROM project GROUP BY state;')

state   total
0    canceled   36659
1      failed  211783
2        live    3163
3      purged     202
4  successful  155130
5   suspended    1700

### How much money has been pledged to campaigns in each country?

In [26]:
dbquery('SELECT country, SUM(usd_pledged) AS total FROM project GROUP BY country ORDER BY total DESC;')

country            total
0       US 3,270,883,676.02
1       GB   298,732,943.23
2       CA   112,432,907.01
3       AU    54,965,224.87
4       DE    54,424,475.45
5       FR    48,536,026.19
6       NL    36,543,906.82
7       IT    25,627,131.64
8       ES    24,281,129.46
9       HK    22,263,452.47
10      SE    20,266,478.73
11      CH    19,435,417.64
12      AT    14,451,653.89
13      DK    11,033,924.53
14      NZ    10,581,166.47
15      SG     9,507,867.91
16      BE     5,963,505.71
17      IE     5,678,333.70
18      JP     4,676,375.59
19      MX     4,091,229.60
20      NO     3,897,921.52
21      LU       546,185.66

## Conversion to CSV `2 points`

### Save the result of your last query - the one about amount raised - to a CSV

In [29]:
dbquery('SELECT country, SUM(usd_pledged) AS total FROM project GROUP BY country ORDER BY total DESC;').to_csv('ranking.csv')

### Visualize the results using DataWrapper, highlighting the country of your choice

Link in the cell below.

[https://datawrapper.dwcdn.net/qy6bY/3/](https://datawrapper.dwcdn.net/qy6bY/3/)

### Building on your last query, calculate the number of projects in each category `2 points`

In [39]:
projects = dbquery('SELECT category.name, COUNT(project.id) AS total \
    FROM project INNER JOIN category on project.category_id = category.id \
    GROUP BY project.category_id')

projects

name  total
0                Art   8725
1             Comics   5194
2              Dance   2389
3             Design   3884
4            Fashion   8303
..               ...    ...
164      Makerspaces    265
165           Comedy     39
166           Comedy    134
167           Comedy    172
168  Literary Spaces     70

[169 rows x 2 columns]

### Using pandas, sort the result to get the top 20 project categories `1 point`

In [40]:
projects.sort_values('total',ascending=False).head(20)

name  total
22     Product Design  24270
9               Music  17055
28     Tabletop Games  16702
24        Documentary  15961
26             Shorts  12572
29        Video Games  12492
5                Food  11054
6        Film & Video  10523
41            Fiction   9510
0                 Art   8725
42         Nonfiction   8623
4             Fashion   8303
65            Apparel   8299
40   Children's Books   7477
134              Apps   7198
12            Theater   7147
37               Rock   6971
11         Technology   6034
27          Webseries   5855
34         Indie Rock   5831